pkg/storage/azureblob/azureblob.go

```diff
@@ -0,0 +1,156 @@
+package azureblob
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"github.com/gomods/athens/pkg/config"
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+)
+
+type client interface {
+	UploadWithContext(ctx context.Context, path, contentType string, content io.Reader) error
+	BlobExists(ctx context.Context, path string) (bool, error)
+	ReadBlob(ctx context.Context, path string) (io.ReadCloser, error)
+	ListBlobs(ctx context.Context, prefix string) ([]string, error)
+	DeleteBlob(ctx context.Context, path string) error
+	GetTimeout() time.Duration
+}
+
+type azureBlobStoreClient struct {
+	containerURL *azblob.ContainerURL
+	timeout      time.Duration
+}
+
+func newBlobStoreClient(accountURL *url.URL, accountName, accountKey, containerName string, timeout time.Duration) (*azureBlobStoreClient, error) {
+	const op errors.Op = "azureblob.newBlobStoreClient"
+	cred, err := azblob.NewSharedKeyCredential(accountName, accountKey)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	pipe := azblob.NewPipeline(cred, azblob.PipelineOptions{})
+	serviceURL := azblob.NewServiceURL(*accountURL, pipe)
+	// rules on container names:
+	// https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names
+	//
+	// This container must exist
+	containerURL := serviceURL.NewContainerURL(containerName)
+	cl := &azureBlobStoreClient{containerURL: &containerURL, timeout: timeout}
+	return cl, nil
+}
+
+// Storage implements (github.com/gomods/athens/pkg/storage).Saver and
+// also provides a function to fetch the location of a module
+type Storage struct {
+	client  client
+	timeout time.Duration
+}
+
+// New creates a new azure blobs storage
+func New(conf *config.AzureBlobConfig, timeout time.Duration) (*Storage, error) {
+	const op errors.Op = "azure.New"
+	u, err := url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net", conf.AccountName))
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	cl, err := newBlobStoreClient(u, conf.AccountName, conf.AccountKey, conf.ContainerName, timeout)
```

Comment:
Do we need the timeout in client if we set it on storage (L 66)? We shouldn't need it in two places.